pushsync/simulation_test.go

```diff
@@ -0,0 +1,269 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the Swarm library. If not, see <http://www.gnu.org/licenses/>.
+
+package pushsync
+
+import (
+	"context"
+	"encoding/hex"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"math/rand"
+	"os"
+	"path/filepath"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/node"
+	"github.com/ethereum/go-ethereum/p2p"
+	"github.com/ethereum/go-ethereum/p2p/simulations/adapters"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/ethersphere/swarm/chunk"
+	"github.com/ethersphere/swarm/log"
+	"github.com/ethersphere/swarm/network"
+	"github.com/ethersphere/swarm/network/retrieval"
+	"github.com/ethersphere/swarm/network/simulation"
+	"github.com/ethersphere/swarm/pss"
+	"github.com/ethersphere/swarm/storage"
+	"github.com/ethersphere/swarm/storage/localstore"
+	"golang.org/x/sync/errgroup"
+)
+
+var (
+	bucketKeyPushSyncer = simulation.BucketKey("pushsyncer")
+	bucketKeyNetStore   = simulation.BucketKey("netstore")
+)
+
+var (
+	nodeCntFlag   = flag.Int("nodes", 4, "number of nodes in simulation")
+	chunkCntFlag  = flag.Int("chunks", 4, "number of chunks per upload in simulation")
+	testCasesFlag = flag.Int("cases", 4, "number of concurrent upload-download cases to test in simulation")
+)
+
+// test syncer using pss
+// the test
+// * creates a simulation with connectivity loaded from a snapshot
+// * for each test case, two nodes are chosen randomly, an uploader and a downloader
+// * uploader uploads a number of chunks
+// * wait until the uploaded chunks are push-synced
+// * downloader has one shot to download all the chunks
+// Testcases are run concurrently
+func TestPushsyncSimulation(t *testing.T) {
+	nodeCnt := *nodeCntFlag
+	chunkCnt := *chunkCntFlag
+	testcases := *testCasesFlag
+
+	err := testPushsyncSimulation(nodeCnt, chunkCnt, testcases, newServiceFunc)
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+func testPushsyncSimulation(nodeCnt, chunkCnt, testcases int, sf simulation.ServiceFunc) error {
+	sim := simulation.NewBzzInProc(map[string]simulation.ServiceFunc{
+		"pushsync": sf,
+	})
+	defer sim.Close()
+
+	ctx := context.Background()
+	snapCtx, cancel := context.WithTimeout(ctx, 30*time.Second)
+	defer cancel()
+	err := sim.UploadSnapshot(snapCtx, filepath.Join("../network/stream/testdata", fmt.Sprintf("snapshot_%d.json", nodeCnt)))
+	if err != nil {
+		return fmt.Errorf("error while loading snapshot: %v", err)
+	}
+
+	start := time.Now()
+	log.Error("Snapshot loaded. Simulation starting", "at", start)
+	result := sim.Run(ctx, func(ctx context.Context, sim *simulation.Simulation) error {
+		var errg errgroup.Group
+		for j := 0; j < testcases; j++ {
+			j := j
```

Comment:
not needed

Reply:
I think that it is needed, as it is used in Go method which calls a provided function in a goroutine.